examples/java-matter-controller/java/src/com/matter/controller/commands/common/Command.java

```diff
@@ -174,16 +189,51 @@ public final int addArgument(String name, StringBuffer out, @Nullable String des
    * @param args Supplied command-line arguments as an array of String objects.
    */
   public final void initArguments(int argc, String[] args) {
-    int argsCount = mArgs.size();
+    int argvExtraArgsCount = argc;
+    int mandatoryArgsCount = 0;
+    int optionalArgsCount = 0;
+
+    for (Argument arg : mArgs) {
+      if (arg.isOptional()) {
+        optionalArgsCount++;
+      } else {
+        mandatoryArgsCount++;
+        argvExtraArgsCount--;
+      }
+    }
 
-    if (argsCount != argc) {
+    if (argc < mandatoryArgsCount) {
       throw new IllegalArgumentException(
-          "Wrong arguments number: " + argc + " instead of " + argsCount);
+          "initArguments: Wrong arguments number: " + argc + " instead of " + mandatoryArgsCount);
     }
 
-    for (int i = 0; i < argsCount; i++) {
+    // Initialize mandatory arguments
+    for (int i = 0; i < mandatoryArgsCount; i++) {
       initArgument(i, args[i]);
     }
+
+    // Initialize optional arguments
+    // Optional arguments expect a name and a value, so i is increased by 2 on every step.
+    for (int i = mandatoryArgsCount; i < argc; i += 2) {
```

Comment:
This code feels tricky ... could we update it to have some different non-nested loop argument paths? 

Specifically:
  - this seems to go to args[i] ... but does this mean that we need to order mandatory arguments before optional? I think arguments should be order-independent right?
  - nested loops are hard to reason about.  See notes below.

Reply:
Looking more at this, terminology is often like this:
  - mandatory arguments are `--<name> ...` that MUST be specified
  - optional arguments are `--<name> ... ` that may or may not be specified
  - positional arguments are based on their position and do not have a `--` or `-` prefix
  
  Could we use that terminology? Could we use an argument parser library to handle this for us?

Reply:
If no other library exists, I would suggest to have a loop that splits out data into:
   - String[] positionalArguments
   - Map<String, String> namedArguments
   
And then do processing based on that data.

Reply:
This logic is ported from chip-tool and looks indeed very tricky. I would suggest to keep aligned with the chip-tool since this will save us a lot of effort to reuse the existing test cases. 

This design is unique, but very intelligent, it allows each command to propose its own argument spec and each argument get constructed during parsing phase, so, none of the existing argument parser can fulfill our requirement since it is not integrated with CHIP command 

Currently, chip-tool require all optional arguments come after mandatory arguments, and all mandatory arguments are positional arguments without -- prefix
Usage:
  ./chip-tool softwarediagnostics read attr-name destination-id endpoint-ids [--paa-trust-store-path] [--cd-trust-store-path] [--commissioner-name] [--commissioner-nodeid] [--use-max-sized-certs] [--only-allow-trusted-cd-keys] [--trace_file] [--trace_log] [--trace_decode] [--ble-adapter] [--fabric-filtered] [--data-version] [--timeout]

All optional arguments are name value pair and its name always start with prefix '--'.

I think I can rewrite this validation logic with single event loop but still follow the same argument spec as chip-tool.